### PR TITLE
don't call user's unmap_dma callback when removing DMA region

### PR DESCRIPTION
**Pull request by tmakatos**
Plus unit tests.

Signed-off-by: Thanos Makatos <thanos.makatos@nutanix.com>
Reported-by: Changpeng Liu <changpeng.liu@intel.com>

### CODE REVIEW

**Review by jlevon — CHANGES_REQUESTED** [review]

File: lib/libvfio-user.c [hunk 1]
```diff
@@ -557,10 +557,10 @@ handle_dma_map_or_unmap(vfu_ctx_t *vfu_ctx, uint32_t size, bool map,
                     dma_regions[i].addr,
                     dma_regions[i].addr + dma_regions[i].size - 1);
         }
-        if (ret < 0) {
+        if (ret < 0) { /* FIXME this if will never be taken */
```
[the remark this answers is not in the document]
Comment (tmakatos) [reply lib/libvfio-user.c:560]: sure will remove it

File: lib/libvfio-user.c [hunk 2]
```diff
@@ -557,10 +557,10 @@ handle_dma_map_or_unmap(vfu_ctx_t *vfu_ctx, uint32_t size, bool map,
...
             return ret;
         }
-        if (vfu_ctx->map_dma != NULL) {
+        if (map && vfu_ctx->map_dma != NULL) {
```
[the remark this answers is not in the document]
Comment (tmakatos) [reply lib/libvfio-user.c:563]: good point

File: test/unit-tests.c [hunk 3]
```diff
@@ -265,6 +273,42 @@ test_dma_map_return_value(void **state UNUSED)
...
+/*
+ * Tests that handle_dma_map_or_unmap calls dma_controller_remove_region.
+ */
+static void
+test_handle_dma_unmap(void **state UNUSED)
+{
+    dma_controller_t d = { 0 };
+    vfu_ctx_t v = {
+        .dma = &d,
+        .map_dma = (void*)0xcafebabe,
```
Comment (jlevon) [inline test/unit-tests.c:285]: nits here, below

void *

**Review by jlevon — APPROVED** [review]
just 3 tiny nits

**Review by swapnili — COMMENTED** [review]
The commit message should say 'map_dma callback'?

**Comment by tmakatos** [comment]
> The commit message should say 'map_dma callback'?

Correct, the commit message is wrong, unfortunately I just merged it :/

**Pull request merged**